# ***BenchMark***

In [ ]:
from reporte_metricas import ReporteMetricas

report = ReporteMetricas()
df = report.load()


Loaded 9 report(s)


## ***Validation***

### ***macro avg***

In [2]:
dfval_macro = df.query("Type == 'val' and Class == 'macro avg'").drop(columns=['accuracy', 'log_loss', 'auc']).reset_index(drop=True)
dfval_macro.sort_values(by=['f1-score'], ascending=False, inplace=True)
dfval_macro

Model Type      Class precision recall f1-score support
6           XGBOOST  val  macro avg      0.86  0.811    0.833  5085.0
8  XGBOOST_Balanced  val  macro avg     0.844  0.815    0.829  5085.0
2              LGBM  val  macro avg     0.849  0.803    0.824  5085.0
1               MLP  val  macro avg     0.807  0.769    0.786  5085.0
0                RF  val  macro avg     0.766  0.792    0.778  5085.0
3               SVC  val  macro avg     0.756  0.744     0.75  5085.0
5                RL  val  macro avg      0.72  0.771     0.74  5085.0
7               KNN  val  macro avg     0.736  0.639    0.673  5085.0
4              CART  val  macro avg     0.649  0.657    0.652  5085.0

### ***weighted avg***

In [3]:
dfval_weight = df.query("Type == 'val' and Class == 'weighted avg'").drop(columns=['accuracy', 'log_loss', 'auc']).reset_index(drop=True)
dfval_weight.sort_values(by=['f1-score'], ascending=False, inplace=True)
dfval_weight

Model Type         Class precision recall f1-score support
6           XGBOOST  val  weighted avg     0.891  0.895    0.892  5085.0
8  XGBOOST_Balanced  val  weighted avg     0.886  0.889    0.887  5085.0
2              LGBM  val  weighted avg     0.884  0.888    0.885  5085.0
1               MLP  val  weighted avg     0.859  0.864    0.861  5085.0
0                RF  val  weighted avg     0.857  0.849    0.852  5085.0
3               SVC  val  weighted avg     0.835  0.838    0.836  5085.0
5                RL  val  weighted avg     0.833  0.815    0.822  5085.0
7               KNN  val  weighted avg     0.794  0.809    0.794  5085.0
4              CART  val  weighted avg     0.761  0.754    0.757  5085.0

### ***Global***

In [4]:
df_val_global = df.query("Type == 'val' and Class == 'global'").drop(columns=['precision', 'recall', 'f1-score', 'support']).reset_index(drop=True)
df_val_global.sort_values(by=['auc'], ascending=False, inplace=True)
df_val_global

Model Type   Class accuracy log_loss    auc
6           XGBOOST  val  global    0.895    0.294  0.967
2              LGBM  val  global    0.888    0.303  0.964
8  XGBOOST_Balanced  val  global    0.889    0.306  0.963
1               MLP  val  global    0.864    0.367   0.95
0                RF  val  global    0.849    0.463  0.949
3               SVC  val  global    0.838    0.419  0.934
5                RL  val  global    0.815    0.561  0.926
7               KNN  val  global    0.809    1.883    0.9
4              CART  val  global    0.754    0.752  0.843

## ***Test***

### ***macro avg***

In [5]:
dfval_macro = df.query("Type == 'test' and Class == 'macro avg'").drop(columns=['accuracy', 'log_loss', 'auc']).reset_index(drop=True)
dfval_macro.sort_values(by=['f1-score'], ascending=False, inplace=True)
dfval_macro

Model  Type      Class precision recall f1-score support
8  XGBOOST_Balanced  test  macro avg      0.85  0.806    0.826  6355.0
6           XGBOOST  test  macro avg     0.844  0.808    0.825  6356.0
2              LGBM  test  macro avg     0.849  0.793    0.818  6355.0
1               MLP  test  macro avg     0.814  0.767    0.788  6355.0
0                RF  test  macro avg     0.771  0.789    0.779  6355.0
3               SVC  test  macro avg     0.756  0.753    0.754  6356.0
5                RL  test  macro avg     0.733  0.786    0.754  6356.0
7               KNN  test  macro avg     0.745  0.643    0.678  6356.0
4              CART  test  macro avg     0.649  0.657    0.653  6356.0

### ***weighted avg***

In [6]:
dfval_weight = df.query("Type == 'test' and Class == 'weighted avg'").drop(columns=['accuracy', 'log_loss', 'auc']).reset_index(drop=True)
dfval_weight.sort_values(by=['f1-score'], ascending=False, inplace=True)
dfval_weight

Model  Type         Class precision recall f1-score support
8  XGBOOST_Balanced  test  weighted avg     0.887  0.891    0.888  6355.0
6           XGBOOST  test  weighted avg     0.885  0.888    0.886  6356.0
2              LGBM  test  weighted avg     0.884  0.889    0.885  6355.0
1               MLP  test  weighted avg     0.861  0.866    0.862  6355.0
0                RF  test  weighted avg     0.856   0.85    0.852  6355.0
3               SVC  test  weighted avg     0.834  0.836    0.835  6356.0
5                RL  test  weighted avg      0.84  0.823    0.829  6356.0
7               KNN  test  weighted avg     0.797  0.811    0.795  6356.0
4              CART  test  weighted avg     0.762  0.757     0.76  6356.0

During the ***benchmarking*** process, various classification models were compared by collecting relevant metrics on both the validation and test sets. The ***unbalanced XGBoost*** model showed the ***best overall performance*** in terms of ***F1-score*** on the validation set, reaching values of ***\$0.833\$ (macro)*** and ***\$0.892\$ (weighted)***, slightly outperforming the model balanced with ADASYN.

On the test set, although the balanced model showed a ***marginal improvement*** of ***\$+0.001\$ to \$+0.002\$*** in the ***F1-score***, these differences are not considered significant from a practical standpoint. Therefore, it is concluded that balancing ***did not substantially improve*** the model's performance and may even introduce bias through artificially generated minority samples via oversampling.

In addition to the aforementioned metrics, it was observed that although the balanced model offered a slight advantage in the 'No Payment' class (***\$+0.01\$***), ***it did not improve performance*** in the minority classes ('Partial Payment' and 'Full Payment'), where the ***F1-score*** was virtually equivalent. This reinforces the claim that balancing did not offer substantial benefits and, furthermore, due to the nature of the organization's operations, this marginal improvement in the majority class provides no practical advantage and unnecessarily increases model complexity.

Another key point is the ***stability of the XGBoost model*** across both sets, maintaining minimal differences between validation and test results, which indicates ***good generalization capacity***. From a practical perspective, synthetic oversampling was discarded as it did not provide relevant improvements, thereby avoiding the potential introduction of artificial noise that could affect the model's interpretability and robustness.

Regarding the comparison with other types of models, there is a ***systematic superiority*** of ***gradient-based*** models, such as ***XGBoost*** and ***LightGBM***, over classical approaches like ***decision trees***, ***logistic regression***, ***SVC***, and ***KNN***, and even over neural networks such as ***MLP***. This advantage is reflected both in average metrics (**macro** and **weighted**) and in consistency across validation and test sets. Moreover, this behavior aligns with recent literature, where *boosting* models have shown excellent capabilities in handling non-linear relationships, imbalanced datasets, and complex classification tasks in the financial and credit risk domains. In industrial settings, their use has also been consolidated due to their ***strong performance, speed, interpretability, and ease of fine-tuning***.

Finally, the model selected as the ***winner*** in the ***benchmarking*** process was the ***unbalanced XGBoost***, prioritizing both its technical performance and adaptability to production environments. Based on this model, the local explanation system using LIME and the integration with the fuzzy inference engine were built, as described in the following sections.